org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/lookup/NestedTypeBinding.java

```diff
@@ -33,7 +33,7 @@ public NestedTypeBinding(char[][] typeName, ClassScope scope, SourceTypeBinding
 	this.enclosingType = enclosingType;
 }
 
-public NestedTypeBinding(NestedTypeBinding prototype) {
+public NestedTypeBinding(NestedTypeBinding prototype, SourceTypeBinding enclosingType) {
```

Comment:
this is where you miss to assign `enclosingType` leading to test regressions.

As the connection is far from obvious: the bug happens below `AnnotatableTypeSystem.getAnnotatedType(TypeBinding, TypeBinding, AnnotationBinding[])` where a pre-computed `enclosingType` with type annotations needs to be set as the enclosing of this nested. Previously, `MemberTypeBinding.clone()` would assign enclosing _after_ creating the clone.